pkg/download/buffer.go

```diff
@@ -65,19 +67,34 @@ type firstReqResult struct {
 	err      error
 }
 
+func readBody(reader *multireader.BufferedReader, resp *http.Response) error {
+	expectedBytes := resp.ContentLength
+	_ = reader.SetSize(expectedBytes)
+	n, err := reader.ReadFrom(resp.Body)
+	if errors.Is(err, io.EOF) {
+		reader.Err = fmt.Errorf("error reading response for %s: %w", resp.Request.URL.String(), err)
+		return reader.Err
+	}
+	if n != expectedBytes {
+		reader.Err = fmt.Errorf("downloaded %d bytes instead of %d for %s", n, expectedBytes, resp.Request.URL.String())
+		return reader.Err
+	}
+	return nil
+}
```

Comment:
I think I must be missing some important context. Why do we need our own custom type/implementation if this is what we're doing with it. This appears to be reading the entire body into a `bytes.Buffer` in one go and then checking the length. At that point how is it different from:

```golang
func readBody(buf *bytes.Buffer, resp *http.Response) error {
	n, err := buf.ReadFrom(resp.Body)
 	if err != nil {
		return err
	}
	if n != resp.ContentLength {
		return fmt.Errorf("downloaded %d bytes instead of %d for %s", n, expectedBytes, resp.Request.URL.String())
	}
	return nil
}
```

(By the by, `BufferedReader.ReadFrom` effectively delegates to `bytes.Buffer.ReadFrom`, which means it cannot return `io.EOF`.)

(Oh, and this code also just completely ignores the case where `err != nil`, which doesn't seem safe.)